library/src/main/java/com/firebase/ui/auth/core/AuthProviderType.java

```diff
@@ -0,0 +1,61 @@
+package com.firebase.ui.auth.core;
+
+import android.content.Context;
+
+import com.firebase.client.Firebase;
+import com.firebase.ui.R;
+import com.firebase.ui.auth.facebook.FacebookAuthProvider;
+import com.firebase.ui.auth.google.GoogleAuthProvider;
+import com.firebase.ui.auth.password.PasswordAuthProvider;
+import com.firebase.ui.auth.twitter.TwitterAuthProvider;
+
+import java.lang.reflect.InvocationTargetException;
+
+public enum AuthProviderType {
+    GOOGLE  ("google",   "com.firebase.ui.auth.google.GoogleAuthProvider",     R.id.google_button),
+    FACEBOOK("facebook", "com.firebase.ui.auth.facebook.FacebookAuthProvider", R.id.facebook_button),
+    TWITTER ("twitter",  "com.firebase.ui.auth.twitter.TwitterAuthProvider",   R.id.twitter_button),
+    PASSWORD("password", "com.firebase.ui.auth.password.PasswordAuthProvider", R.id.password_button);
+
+    private final String mName;
+    private final String mProviderName;
+    private final int mButtonId;
+
+    AuthProviderType(String name, String providerName, int button_id) {
+        this.mName = name;
+        this.mProviderName = providerName;
+        this.mButtonId = button_id;
+    }
+
+    public String getName() {
+        return mName;
+    }
+    public int getButtonId() {
+        return mButtonId;
+    }
+
+    public FirebaseAuthProvider createProvider(Context context, Firebase ref, TokenAuthHandler handler) {
+        try {
+            Class<? extends FirebaseAuthProvider> clazz = (Class<? extends FirebaseAuthProvider>) Class.forName(mProviderName);
```

Comment:
There is definitely a [convention to call these clazz](http://stackoverflow.com/questions/2529974/why-do-java-programmers-like-to-name-a-variable-clazz), but I'm open to any other name.